### ISSUE TITLE
Compare Unkown jsonstring with Nested arrays  

### ISSUE THREAD

**Issue by Bishalsahoo** [Tencent/rapidjson#1224]
Hello , I am  a bit stuck in comparing two json strings

`#include <iostream>
#include "include/rapidjson/document.h"
#include "include/rapidjson/writer.h"
//#include "include/rapidjson/stringbuffer.h"
using namespace std;
using namespace rapidjson;
class test {
public:
    static bool isequalstring(const std::string &item1, const std::string &item2, const std::string &temp) {
//d1,d2,d3 document type variable to parse the provided string
        Document d1;
        d1.Parse(item1.c_str());
        Document d2;
        d2.Parse(item2.c_str());

        Document d3;
        d3.Parse(temp.c_str());
        StringBuffer buffer;
        Writer<StringBuffer> writer(buffer);
        d1.Accept(writer);
        StringBuffer buffer1;
        Writer<StringBuffer> writer1(buffer);
        d2.Accept(writer1);
        StringBuffer buffer2;
        Writer<StringBuffer> writer2(buffer);
        d3.Accept(writer2);

        // Output {"project":"rapidjson","stars":11}
        std::cout << buffer.GetString() << std::endl;
        std::cout << buffer1.GetString() << std::endl;
        std::cout << buffer2.GetString() << std::endl;

        bool matched = true;
        //itr= iterate through the third json to get the keys and match the keys in first and second
        for (Value::ConstMemberIterator itr = d3.MemberBegin(); itr != d3.MemberEnd(); itr++) {

                if (d1.HasMember(itr->name) && d2.HasMember(itr->name)) {
                                                                            // if the member doesn't exist in both, break
                    if (d1[itr->name] != d2[itr->name]) {

                    string n = itr->name.GetString();

cout<<n;
                        // value doesn't match, then break
                        matched = false;

                    }

                } else {
                    matched = false;

                    break;
                }
            }

        return matched;
    }
};

bool testDeepNestedJson_should_succeed(){
    bool expectedTestResult = true;
    bool testResult;

  
    // Input 1 JSON Object
    const char* input1 = "{\"array\":[1,2,3],\"boolean\":true,\"null\":null,\"number\":123,\"object\":{\"a\":\"b\",\"c\":\"d\",\"e\":\"f\"},\"string\":\"Hello World\",\"object_array\":[{\"key\":\"value1\"},{\"key\":\"value2\"},{\"key\":\"value3\"}],\"deep_nested_array\":[{\"object_array2\":[{\"key\":\"value1\"},{\"key\":\"value2\"},{\"key\":\"value3\"}]},{\"object_array1\":[{\"key\":\"value4\"},{\"key\":\"value5\"},{\"key\":\"value6\"}]}]}";
    
    //Input 2 JSON Object
    const char* input2 = "{\"array\":[1,2,3],\"justsomedata\":true,\"boolean\":true,\"null\":null,\"object\":{\"a\":\"b\",\"c\":\"d\",\"e\":\"f\"},\"number\":123,\"object_array\":[{\"whatever\":\"test\",\"key\":\"value1\"},{\"key\":\"value2\"},{\"key\":\"value3\"}],\"deep_nested_array\":[{\"object_array2\":[{\"key\":\"value1\"},{\"key\":\"value2\"},{\"key\":\"value3\"}]},{\"object_array1\":[{\"key\":\"value6\"},{\"key\":\"value4\"},{\"key\":\"value5\",\"ignoreme\":12346}]}],\"string\":\"Hello World\"}";
    
    //Input Template on whose keys the value in the input 1 and input2 will be compared if they are equal or not (its like a reference for comparing the input1 and input2 JSONs
    const char* stencil = "{\"array\":[null],\"boolean\":null,\"null\":null,\"object\":{\"a\":null,\"c\":null,\"e\":null},\"number\":null,\"object_array\":[{\"key\":null}],\"deep_nested_array\":[{\"object_array2\":[{\"key\":null},{\"key\":null},{\"key\":null}]},{\"object_array1\":[{\"key\":null},{\"key\":null},{\"key\":null}]}],\"string\":null}";
   
   
   
   testResult = test::isequalstring(input1, input2,stencil);

    if(testResult != expectedTestResult){
        //std::cout<<"testDeepNestedJson_should_succeed:"<<std::endl;
        //std::cout<<"Item1:"<<input1<<std::endl;
        //std::cout<<"Item2:"<<input2<<std::endl;
        // std::cout<<"Stencil:"<<stencil<<std::endl;

        std::cout<<"Test Failed result is: %%REPLACE WITH testResult%% expected was: %%REPLACE WITH expectedTestResult%%"<<std::endl;

        return false;
    }

    std::cout<<"PASSED: testDeepNestedJson_should_succeed"<<std::endl;

    return true;
}
int main() {
testDeepNestedJson_should_succeed();
    return 0;
}` 

Here I am able to compare the simple Json with single array but when there are n number of nested array it doesnt works the way it should be . I need it to loop into array element and check them individually as there might be sub array . I tried to do that with assert(d1[itr->name.IsArray()]) but it will be a same problem if it has more subarrays in those sub array cant write loops to find if the key is an arrary till n number of times and if there is any other functionality in rapidjson I am unknown as I am absolutely new to rapidjson. Can anyone please help me complete the code to make it effective ? 
In the above case the object_array and deep_nested_array are the one that mismatch even though they are same according to the stencil(the reference) . The output should be true but it is false instead. Please help me with this.

**Comment by Bishalsahoo** [comment]
@miloyip can you help me please ? Its kinda urgent . Hope for your response soon

**Comment by Bishalsahoo** [comment]
i want to compare the keys that only exists in the reference template(stencil)

**Comment by Bishalsahoo** [comment]
https://drive.google.com/open?id=1HY-jfMId7qELNZdm2Iu1tXP-QZr2e-9y please have a look into the raw json. My code is not diving deep enough to get and compare array inside deep_nested_array

**Comment by miloyip** [comment]
I think that you want to **validate a JSON** to ensure it conforms to your "template".
RapidJSON supports JSON Schema which can do it simply by defining a JSON schema.
You may check about what is JSON schema (e.g. [here](https://spacetelescope.github.io/understanding-json-schema/)), and then figure how to use it in RapidJSON in [here](http://rapidjson.org/md_doc_schema.html).

**Comment by Bishalsahoo** [comment]
Yes I want to validate values of two json based on the template and I will be going through that but your help to complete the code will be very helpful as it is kinda urgent please@miloyip

**Comment by miloyip** [comment]
I cannot help finishing your own work. Please understand.

**Comment by Bishalsahoo** [comment]
No I know that but can you show me the way to iterate into nested array get the elements and check if the value in the array is another array and loop in its value to check if they are equal , just need help with this logic as i am able to enter the array (for instance deep_nested_array in this case )but unable to iterate thought the sub array and check the value. whatever i try is showing error . Just help me with the logic I will be able to complete it

**Comment by Bishalsahoo** [comment]
I tried to use the schema too but the problem is the parser is unable to detect the sub nested array and just returns the key on the outer loop , because of that it compares the value of deep_nested_array rather than going deep into it to get the object_array and the key elements inside it. I need a way to get the keys inside the deep_nested_array and then inside the object_array. As I mentioned i tried to loop inside the deep_nested_array and get the index numbers but unable to access the sub element object in it and loop inside them(ie i can loop inside the 1st array but unable to get value or key to compare and loop further with in them). Please help me just with the logic, how to do it.

**Comment by Bishalsahoo** [comment]
Can anyone help me to complete the correct way of recursion by passing the value parameter back to continue the loop for the deeper array
